### PR TITLE
Refactor reports/metrics, add JSON and markdown report

### PR DESCRIPTION
**Pull request by ctron**
This PR refactors the code of the metrics and report module a bit. It extracts the common part into a common module, allowing to prepare the data consistently and then rendering it out into HTML, and also serializing the data into JSON.

For that the `--report-file` argument has been changed into a `Vec<String>` to allow providing more than one file. It uses the file extension as an indicator of which file type should be created.

In addition, this should make it easier to implement a markdown report too.

### CODE REVIEW

**Comment by ctron** [comment]
I updated the PR to make clippy happy.

Although these are not files touched by the PR, it looks like newer Rust versions brought in new lints that now fail.

**Comment by ctron** [comment]
I updated the PR to support markdown as well. It should be pretty close to the HTML report, but without the charts.

**Comment by jeremyandrews** [comment]
This looks great from a read through on my phone. I’m traveling internationally but will find time to test it and get it merged. 

In a follow up PR, it would be great to update the book documentation too.

**Comment by ctron** [comment]
> This looks great from a read through on my phone. I’m traveling internationally but will find time to test it and get it merged.
> 
> In a follow up PR, it would be great to update the book documentation too.

Cool. Take your time. I can update the PR over time. I am currently also trying to add the baseline feature. We can split that off, but it's based on the other changes anyway.

**Comment by JimFuller-RedHat** [comment]
does this mean a new release as well ? nudge nudge

**Comment by ctron** [comment]
I pushed the change for baseline comparison too. Of course we can split this off. Looks like this:

image

**Comment by ctron** [comment]
> > I pushed the change for baseline comparison too. Of course we can split this off.
> 
> Please split this off -- I'd like to get the first bits committed first.

I split it off (#602) and rebase the PR.

**Comment by jeremyandrews** [comment]
I'm finally starting to test this now. We need to improve the documentation, as it's non-obvious what format of report files are supported. Further, it feels like there should be some sort of validation? For example, if you accidentally type ` --report-file report.htm` it will create the file but not log anything there. While this isn't the expected filename on a Unix-based system, wouldn't it be legit on Windows?  And, if you type `--report-file report.foo` it will happily create that filename, but then it won't write data to it.

File: Cargo.toml [hunk 1]
```diff
@@ -52,6 +52,7 @@ url = "2"
 [features]
 default = ["reqwest/default-tls"]
 rustls-tls = ["reqwest/rustls-tls", "tokio-tungstenite/rustls"]
+gaggle = []
```
Comment (jeremyandrews) [inline Cargo.toml:55]: Gaggles are currently not supported in this version of Goose.

File: src/report/markdown.rs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        Ok(())
+    }
+
+    fn write_response_metrics(&mut self) -> Result<(), GooseError> {
+        write!(
+            self.w,
+            r#"
+## Response Time Metrics
+
+| Method | Name | 50%ile (ms) | 50%ile (ms) | 60%ile (ms) | 70%ile (ms) | 70%ile (ms) | 80%ile (ms) | 90%ile (ms) | 95%ile (ms) | 99%ile (ms) | 100%ile (ms) |
```
Comment (jeremyandrews) [inline src/report/markdown.rs:143]: `50%ile (ms)` and `70%ile (ms)` are repeated twice

**Review by jeremyandrews — CHANGES_REQUESTED** [review]
Overall this is fantastic, thanks for the contribution! However, there are a few small issues to resolve, see the earlier comments. In addition, it would be very helpful to:

- Add an entry to CHANGELOG.md
- Update affected documentation, including: `src/docs/goose-book/src/getting-started/running.md` (https://book.goose.rs/getting-started/running.html), `src/docs/goose-book/src/getting-started/runtime-options.md` (https://book.goose.rs/getting-started/runtime-options.html), `src/docs/goose-book/src/getting-started/common.md` (https://book.goose.rs/getting-started/common.html#writing-an-html-formatted-report), `src/docs/goose-book/src/getting-started/metrics.md` (https://book.goose.rs/getting-started/metrics.html), `src/docs/goose-book/src/config/defaults.md` (https://book.goose.rs/config/defaults.html)

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ctron) [reply Cargo.toml:55]: Not sure why that was there. Removing it works now.

File: src/metrics.rs [hunk 3]
```diff
@@ -2994,46 +3003,112 @@ impl GooseAttack {
...
+                Some("html") => {
+                    self.write_html_report(create(path).await?, report).await?;
+                }
+                Some("json") => {
+                    self.write_json_report(create(path).await?).await?;
+                }
+                Some("md") => {
+                    self.write_markdown_report(create(path).await?).await?;
+                }
+                None => {
```
[the remark this answers is not in the document]
Comment (ctron) [reply src/metrics.rs:3030]: It kind of does. However there's a section upfront which creates all file unconditionally to ensure it has access. I'll take a look.

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ctron) [reply Cargo.toml:55]: Ah here, that was it:

```
warning: unexpected `cfg` condition value: `gaggle`
   --> tests/one_taskset.rs:243:16
    |
243 | #[cfg_attr(not(feature = "gaggle"), ignore)]
    |                ^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `default` and `rustls-tls`
    = help: consider adding `gaggle` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration
    = note: `#[warn(unexpected_cfgs)]` on by default
```

File: src/metrics.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ctron) [reply src/metrics.rs:3042]: Ok, fixed it.

File: src/metrics.rs [hunk 4]
```diff
@@ -2994,46 +3003,112 @@ impl GooseAttack {
...
+                    return Err(GooseError::InvalidOption {
+                        option: "--report-file".to_string(),
+                        value: report.clone(),
+                        detail: "Missing file extension for report".to_string(),
+                    })
+                }
+                Some(ext) => {
```
[the remark this answers is not in the document]
Comment (ctron) [reply src/metrics.rs:3049]: Same here, fixed.

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ctron) [reply Cargo.toml]: I removed it anyway.

File: Cargo.toml [hunk 5]
```diff
@@ -52,7 +52,6 @@ url = "2"
 [features]
 default = ["reqwest/default-tls"]
 rustls-tls = ["reqwest/rustls-tls", "tokio-tungstenite/rustls"]
-gaggle = []
```
Comment (jeremyandrews) [inline Cargo.toml]: It seems we're making reference to this feature in the code still, so this is actually required. Sorry that I missed that.

Comment (ctron) [reply Cargo.toml]: Ok, dropped it.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,444 +1,624 @@
...
+
+- [#565](https://github.com/tag1consulting/goose/pull/565) add `--accept-invalid-certs` to skip validation of https
+  certificates
+- [#568](https://github.com/tag1consulting/goose/pull/568) don't panic when truncating non utf-8 string
+- [#574](https://github.com/tag1consulting/goose/pull/574)
+  update [`http`](https://docs.rs/http), [`itertools`](https://docs.rs/itertools) [`nix`](https://docs.rs/nix), [`rustls`](https://docs.rs/rustls/),
+  and [`serial_test`](https://docs.rs/serial_test)
+- [#575](https://github.com/tag1consulting/goose/pull/575) add test coverage for sessions and cookies,
+  revert [#557](https://github.com/tag1consulting/goose/pull/557) to avoid sharing the CookieJar between all users
+- [#600](https://github.com/tag1consulting/goose/pull/600) Refactor reports/metrics, add JSON and markdown report
```
Comment (jeremyandrews) [inline CHANGELOG.md:13]: I think your editor is reformatting this entire file, I'd prefer to not make unrelated changes like this.

Comment (ctron) [reply CHANGELOG.md]: Should be fixed.

Comment (jeremyandrews) [reply CHANGELOG.md]: Perfect, thank you! I've started reviewing, will aim to get it merged quickly (probably tomorrow at this rate).

**Review by jeremyandrews — APPROVED** [review]
This looks great, thanks for all your updates! I'm only confused about the test failures: locally I can't duplicate this, and these tests that are failing should be ignored. Any ideas?

**Comment by jeremyandrews** [comment]
> This looks great, thanks for all your updates! I'm only confused about the test failures: locally I can't duplicate this, and these tests that are failing should be ignored. Any ideas?

I see, we're running with `--all-features`, so adding back the gaggle feature is causing these tests to run, while the contained code is no longer supported (for now).

**Comment by jeremyandrews** [comment]
We can easily fix the gaggle issue in a follow-up, I've confirmed all other tests run correctly and without errors.

This new functionality is greatly appreciated, thanks for the contribution!
